SPA版メモアプリを作成しました

rira100000000 opened the pull request:
[Create React App](https://bootcamp.fjord.jp/practices/235)の提出物です。

rira100000000 committed:
SPA版メモアプリを作成しました

rira100000000 committed:
不要なファイルを削除しました

rira100000000 committed:
不要なimportを削除しました

rira100000000 committed:
eslint対応

rira100000000 committed:
idsを数値の配列に変更

rira100000000 committed:
不要なコメントの削除

rira100000000 committed:
新規メモのID計算処理をuseEffectを使ってリファクタリングしました

rira100000000 committed:
カスタムフックの導入

新規作成と編集ボタンボタンの処理を共通化しました。

m0nch1 reviewed (commented):
@rira100000000 
大小含めて色々コメントさせてもらいました！
（メンター陣からのレビューが遅れてしまい申し訳ないです 🙏 ）

ソースコードのレビューは以下に記載していますが、気を付けると良さそうなことをお伝えします！（次回からで大丈夫です！）

レビューで言っていることがわからない、進め方がわからない、など不安になったら Discord なりペアプロなりしますので気軽にご連絡くださいな！ 👍 

## commit の粒度について

今回、`npx create-react-app my-app` のようなコマンドを用いてアプリケーションを作成しているかと思うのですが、一度その段階で commit を積んでもらえるとレビューがしやすくて助かります！ ☺️ 

※ 今回で言うと https://github.com/rira100000000/react_practice/pull/1/commits/873f0b4d85e1ab1bb80c711f4368914e781ce817 のコミットに初期インストールと実装が混じっている

## Pull Request の Description について
<img width="870" alt="image" src="https://github.com/rira100000000/react_practice/assets/26811986/d407a6ea-3e2a-4210-aa46-5427ba50caca">

ベストエフォートではありますが例えば、「やったこと」、「やらなかったこと」、「レビューして欲しいところ（不安なところ）」、「自分で動作確認したこと」など記載しておくとレビューのときに助かったり、自分で後で PR をみた時も役だったりします！

rira100000000 committed:
jsx構文を使っているファイル拡張子をjsxに修正しました

rira100000000 commented on src/DeleteButton.js:
```diff
@@ -0,0 +1,25 @@
+const DeleteButton = (props) => {
```
jsx構文を使っているファイルは`jsx`に修正します。

rira100000000 committed:
言語設定を日本語に修正しました

rira100000000 commented on public/index.html:
```diff
@@ -0,0 +1,43 @@
+<!DOCTYPE html>
+<html lang="en">
```
jaに修正します。

rira100000000 committed:
localStorageのデータ設計の変更

valueないでオブジェクトとしてデータを持つように修正しました

rira100000000 committed:
Todo.jsxのリファクタリング

if文を||演算子に修正しました。
三項演算子を使って記述するように修正しました。

rira100000000 committed:
contentsの初期化処理をuseEffectで行うように修正しました

rira100000000 commented on src/Todo.js:
```diff
@@ -0,0 +1,88 @@
+import EditForm from "./EditForm";
+
+const Todo = () => {
+  const ids = Object.keys(localStorage)
+    .filter((id) => id !== "id")
+    .map((id) => {
+      return parseInt(id);
+    })
+    .sort((a, b) => a - b);
+
+  const [contents, setContents] = useState(() => {
+    let result = {};
+
+    ids.forEach((id) => {
+      result[id] = localStorage.getItem(id);
+    });
+
+    return result;
+  });
```
valueにオブジェクトを入れる形に変更します。
contentsの初期化処理をuseEffectを使って行うようにします。

rira100000000 committed:
localStorage関係の処理を切り出す

localstorageでのデータ取得、入力処理をuseLocalStorageフックに切り出しました。

rira100000000 commented on src/useSaveButton.js:
```diff
@@ -0,0 +1,17 @@
+const useSaveButton = (props, id, content) => {
+  const { setContent, setContents, setEditingId, setShowEditForm } = props;
+
+  const saveData = React.useCallback(() => {
+    setContent(content);
+    localStorage.setItem(id, content);
+    setContents((prevContents) => ({ ...prevContents, [id]: content }));
+    setEditingId(id);
+    setShowEditForm(true);
+  }, [setContent, setContents, setEditingId, setShowEditForm, id, content]);
```
localStorageの処理`useLocalStorage `として切り出します。

rira100000000 committed:
削除処理でもuseSaveButtonフックを利用するよう修正しました

rira100000000 commented on src/Todo.js:
```diff
@@ -0,0 +1,88 @@
+    return result;
+  });
+
+  const [newId, setNewId] = useState("");
+  useEffect(() => {
+    const calcNewId = () => {
+      if (ids.length === 0) {
+        return 0;
+      } else {
+        return Math.max(...ids) + 1;
+      }
+    };
+    setNewId(calcNewId());
+  }, [ids]);
```
確かに必要ありませんでした🙏
NewButtonで計算するようにします。

rira100000000 committed:
不要なstateを削除

newIdは不要なstateだったため削除し、NewButtonで計算するように修正しました。

rira100000000 commented on src/DeleteButton.js:
```diff
@@ -0,0 +1,25 @@
+    <div>
+      <button
+        className="delete"
+        id={"delete" + props.editingId}
```
テンプレートリテラルを使うよう修正します！

rira100000000 committed:
文字列の作成にテンプレートリテラルを使用するようにしました

rira100000000 commented on src/ShowButton.js:
```diff
@@ -0,0 +1,27 @@
+  return (
+    <div>
+      <button
+        id={"show_" + props.id}
```
テンプレートリテラルを使うよう修正します。
deleteに`_`を付けて命名規則を一致させます。

rira100000000 committed:
受け渡すpropsの見直し

NewButton,EditForm,DeleteButtonに渡すpropsを見直しました。
todoに対する処理をTodo.jsxに定義して各コンポーネントに渡すようにしました。

rira100000000 committed:
showButtonのpropsの見直し

handleclickの処理をTodoに実装してpropsとして渡すようにしました。
不要なpropsを削除しました。

rira100000000 commented on src/Todo.js:
```diff
@@ -0,0 +1,88 @@
+            return (
+              <div key={"ShowButton_" + id}>
+                <ShowButton
+                  id={id}
+                  contents={contents}
+                  setContents={setContents}
+                  setShowEditForm={setShowEditForm}
+                  editingId={editingId}
+                  setEditingId={setEditingId}
+                  setContent={setContent}
```
一覧表示、新規作成、更新、削除の処理を関数として渡すようにし、不要になったpropsを削除します。

rira100000000 commented on src/Todo.js:
```diff
@@ -0,0 +1,88 @@
+            );
+          })}
+          <NewButton
+            newId={newId}
+            contents={contents}
+            setContents={setContents}
+            setContent={setContent}
+            setShowEditForm={setShowEditForm}
+            setEditingId={setEditingId}
```
一覧表示、新規作成、更新、削除の処理を関数として渡すようにし、不要になったpropsを削除します。

rira100000000 commented on src/Todo.js:
```diff
@@ -0,0 +1,88 @@
+        <div className="contentArea">
+          {showEditForm && (
+            <EditForm
+              editingId={editingId}
+              setEditingId={setEditingId}
+              contents={contents}
+              setContents={setContents}
+              content={content}
+              setContent={setContent}
+              setShowEditForm={setShowEditForm}
```
一覧表示、新規作成、更新、削除の処理を関数として渡すようにし、不要になったpropsを削除します。

rira100000000 committed:
CSSの修正

ボタンにマウスオンでカーソルが変更されるようにしました。
不要なCSSを削除しました。

rira100000000 commented on src/index.css:
```diff
@@ -0,0 +1,71 @@
+}
+
+.delete {
+  border-radius: 5px;
```
良い感じになりました😄
各ボタンに`cursor: pointer;`を追加します。

rira100000000 commented on src/index.css:
```diff
@@ -0,0 +1,71 @@
+  -moz-osx-font-smoothing: grayscale;
+}
+
+code {
```
不要でしたので削除します。

rira100000000 committed:
軽微な修正

idsの寿命が短くなるよう定義する位置を移動しました。
contentの初期値は利用されないので空の文字列に修正しました。

rira100000000 committed:
console.logを削除しました。

rira100000000 committed:
不要なimportを削除しました。

rira100000000 commented on src/Todo.js:
```diff
@@ -0,0 +1,88 @@
+
+  const [showEditForm, setShowEditForm] = useState(false);
+  const [editingId, setEditingId] = useState("");
+  const [content, setContent] = useState(contents[editingId]);
```
すいません、ここだけやり方がわかりませんでした🙇‍♂️

`Todo`に移された処理でもこれらのデータを使いたいので、`EditForm`に移すと子コンポーネントから親が読み取る形になってしまうと思うのですが、他の手段があるのでしょうか？

rira100000000 commented:
1点質問なのですが、`useSaveButton`は`NewButton`と`EditForm`のsubmitの機能を共通化しています。
そのためやたらと長い依存関係配列を持つことになっていたり、一方の処理では不要な処理も行っていて、コードの質が悪い気がしています。

実装当初は「共通化できたらイケてるじゃん！」と思っていたのですが、レビュー修正後のコードでここだけ異質な感じがしていて、素直に分けて書いた方がいいのでは？と思っているのですがどうでしょうか？

---
追記
すいません、もう一点気になってる部分がありました。
Todo.jsxの`useEffect`を使って`setContents`を初期化している部分なのですが、このように実装しています。
```javascript
  useEffect(() => {
    const todos = JSON.parse(localStorage.getItem("todos")) || {};
    setContents(todos);
  }, []);
```
ここのtodosの取得処理は`useLocalStorage`の`getLocalStorageContents`が使えそうなんですが、`getLocalStorageContents`を使いつつ無限ループを引き起こさない方法がわかりませんでした。こちらについても何かヒントをいただきたいです🙇‍♂️

m0nch1 commented:
修正ありがとうございます！
以前よりだいぶみやすくなった気がします 👍 （サイコウサイコウ）

まずはコメントいただいていたところについて回答しますねー！

> 1点質問なのですが、useSaveButtonはNewButtonとEditFormのsubmitの機能を共通化しています。
そのためやたらと長い依存関係配列を持つことになっていたり、一方の処理では不要な処理も行っていて、コードの質が悪い気がしています。

>実装当初は「共通化できたらイケてるじゃん！」と思っていたのですが、レビュー修正後のコードでここだけ異質な感じがしていて、素直に分けて書いた方がいいのでは？と思っているのですがどうでしょうか？

そうですねー、無理に DRY (Don't Repeat Yourself) にする必要はないと思います！
どんな処理を hooks にすべきか？と言う観点で考えてみるといいのかもしれません！
依存関係配列が多いとどうなの？については以下あたりを読んでみるとヒントが得られるかと
refs: 
- https://legacy.reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often
- https://overreacted.io/ja/a-complete-guide-to-useeffect/

現状の `useSaveButton` は以下の様々な操作をしていますが、命名からしてももっとスリムな hooks にするべきな気がしますね！
- content を設定する
- 最新のコンテンツにする
- 編集中の id をセットする
- 編集フォームの表示状態を変更する
- ローカルストレージへデータを同期する

やりたいことは以下のようなイメージだと思うんですよね... 🤔 

-  + ボタンを押下して新しい TODO を追加したとき
  - state に新しいオブジェクトを追加する
  - 追加した後、localstorage にもデータを同期する
- 編集ボタンを押下して既存のデータを更新したとき
  - state の対象のデータを更新する
  - 更新した後、localstorage にもデータを同期する

このように考えると、props に渡すものなどまた考える必要は出てきますが、
contents に対しての操作を、ローカルストレージの操作だけで済みそうな感じがしてきましたね！
詳しくはまたコードにコメントしていきますねー！

---

> ここのtodosの取得処理はuseLocalStorageのgetLocalStorageContentsが使えそうなんですが、getLocalStorageContentsを使いつつ無限ループを引き起こさない方法がわかりませんでした。こちらについても何かヒントをいただきたいです🙇‍♂️

順を追ってみていくと意外とシンプルな理由です！

- useEffect の中で getLocalStorageContents を実行
- getLocalStorageContents でローカルストレージからデータを取得し、そのデータを setContents で contents ステートにセット
- setContents をしたので contents ステートが更新されるが、contents は useCallback の依存配列にあるためgetLocalStorageContents が再度実行される。

これが繰り返されて無限ループになっています！

今回の場合、state の更新は hooks を使う側のコンポーネントに任せてこの hooks ではローカルストレージのデータを読み書きすることに絞った方が良さそうです！

イメージはこんな感じ

```Todo.jsx
  useEffect(() => {
    const todos = getLocalStorageContents();
    setContents(todos);
  }, []);
```

```useLocalStorage.js
import { useCallback } from "react";

const useLocalStorage = () => {
  const getLocalStorageContents = useCallback(() => {
    const storedContents = localStorage.getItem("todos");
    return storedContents ? JSON.parse(storedContents) : [];
  }, []);

  const syncContentsToLocalStorage = (newContents) => {
    localStorage.setItem("todos", JSON.stringify([newContents]));
  };

  return {
    syncContentsToLocalStorage, 
    getLocalStorageContents
  };
};

export default useLocalStorage;

```

m0nch1 commented on src/useSaveButton.js:
```diff
@@ -0,0 +1,30 @@
+import { useCallback } from "react";
```
問題のこの hooks ですが、僕がコメントに書いたイメージをざっくり書きますね！
（動作確認してないので動くかわからないコードです。あくまで参考程度に）

名前も変えてしまっているんですが、こんな形で state も hooks に持たせてあげるとやりたいことができるのではないかなーと思いました

コードの例では削除などはしてませんが、deleteContent など作って中身では id を props に検索してきたものを分割代入などで無くしてしまって、それを setContents すると delete 構文も無くなってより良いかなと思います 👍 


```
import { useState, useEffect, useMemo } from "react";
import useLocalStorage from "./useLocalStorage";

const useContents = () => {
  const [contents, setContents] = useState([]);
  const {syncContentsToLocalStorage, getLocalStorageContents} = useLocalStorage();

  useEffect(() => {
    const todos = getLocalStorageContents()
    setContents(todos);
  }, [getLocalStorageContents]);

  const initialContents = {
    id: contents.length + 1,
    text: "新規メモ",
  };

  const addContent = () => {
    const newContents = [...contents, initialContents];
    setContents(newContents)
    syncContentsToLocalStorage(newContents)
  }

  const updateContent = (contentsId, text) => {
    const updatedContents = contents.map(content => content.id === contentsId ? {...content, content: text} : content);
    setContents(updatedContents)
    syncContentsToLocalStorage(updatedContents)
  }

  const contentsMemoized = useMemo(() => contents, [contents]);

  return {
    contents: contentsMemoized,
    setContents,
    addContent,
    updateContent
  }
};

export default useContents;
```

m0nch1 commented on src/useLocalStorage.js:
```diff
@@ -0,0 +1,22 @@
+import { useCallback } from "react";
+
+const useLocalStorage = (contents, setContents) => {
+  const getLocalStorageContents = useCallback(() => {
+    const storedContents = localStorage.getItem("todos");
+    if (storedContents) {
+      setContents(JSON.parse(storedContents));
+      return contents;
+    } else {
+      return {};
+    }
+  }, [contents, setContents]);
+
+  const syncContentsToLocalStorage = (newContents) => {
+    localStorage.setItem("todos", JSON.stringify(newContents));
+    setContents(newContents);
+  };
```
ここらは https://github.com/rira100000000/react_practice/pull/1#issuecomment-1609802349 に書いたとおり、localstorage の操作のみにとどめた方が良さそう

m0nch1 commented on src/Todo.jsx:
```diff
@@ -0,0 +1,90 @@
+    return parseInt(id);
+  });
+
+  const saveNewTodo = useSaveButton(
+    { setContent, setContents, setEditingId, setShowEditForm },
+    calcNewId(ids),
+    "新規メモ",
+    contents
+  );
+
+  const updateTodo = useSaveButton(
+    { setContent, setContents, setEditingId, setShowEditForm },
+    editingId,
+    content,
+    contents
+  );
+
+  const deleteTodo = (id) => {
+    let updatedContents = { ...contents };
+    delete updatedContents[id];
+    setContents(updatedContents);
+    syncContentsToLocalStorage(updatedContents);
+    setShowEditForm(false);
+    setEditingId("");
+  };
```
この辺の処理をバクっと hooks に持っていくと気持ちがよさそうと言うイメージ 👍

m0nch1 commented on src/NewButton.jsx:
```diff
@@ -0,0 +1,13 @@
+  );
+};
+
+export const calcNewId = (ids) => {
+  return ids.length === 0 ? 0 : Math.max(...ids) + 1;
+};
```
これ、サンプルのコードにも載せたんですが、`contents.length + 1 ` とかでよさそうですかね？

m0nch1 reviewed (commented):
@rira100000000 
再度コメントさせていただきましたー！
また多めのコメントになってしまったので少しづつ進めてもらえればと！
commit の粒度など細かくてみやすかったので助かってます 👍 

+ １点ここでもおすすめをお伝えすると、src 配下を `components`, `hooks` とかでディレクトリ分けてあげてもいいかもしれません！

![image](https://github.com/rira100000000/react_practice/assets/26811986/8a7e30a6-3b20-4fcc-bca4-1a4eef555580)

rira100000000 committed:
ディレクトリ構成の変更

src配下にcomponents,hooksディレクトリを追加しました

rira100000000 committed:
contentsのデータ構造をオブジェクトから配列に変更

配列contentsにオブジェクトcontentが入り、contentはidとtextを持つ形に変更しました。
todoの本文をcontentと命名していましたが、textに変更しました。

rira100000000 commented on src/Todo.jsx:
```diff
@@ -0,0 +1,90 @@
+import useSaveButton from "./useSaveButton";
+
+const Todo = () => {
+  const [contents, setContents] = useState({});
```
配列contentsの中でオブジェクトcontentを持つようにします。
オブジェクトcontentはidと本文であるtextを持つようにします。

rira100000000 committed:
useLocalStorageの戻り値を配列からオブジェクトに変更しました

rira100000000 committed:
contentsの操作をuseContentsフックで行うよう修正

useSaveButtonをuseContentsにリネームしました。
Todoに実装していたcontentsの操作をuseContentsフックで行うようにしました。

rira100000000 committed:
軽微な修正

setShowEditForm,setEditingId,setTextの処理の順番を統一しました。
localStorage.Clearを削除しました。
propsに渡すtextの名前を修正しました。

rira100000000 committed:
handleSubmitに渡していた不要な引数を削除しました

rira100000000 committed:
新規登録時の初期テキストをDRYになるよう修正しました。

m0nch1 reviewed (approved):
修正ありがとうございましたー！
LGTM です 👍
